### ISSUE TITLE
Fix pycsw harvesting routine

### ISSUE THREAD

**Issue by dnowacki-usgs** [USGS-CMG/usgs-cmg-portal#313]
The cron jobs were and are failing for pycsw. There are several issues. Biggest one right now is that there are too many records for a clean wipe every day. Propose only wiping those records that change routinely (e.g. model runs) and leave the ts records intact except for when adding new datasets.

**Comment by rsignell-usgs** [comment]
@dnowacki-usgs , that sounds like a great plan!

**Comment by dnowacki-usgs** [comment]
@rsignell-usgs 👍 can you send me the environment definition for IOOS2?

**Comment by rsignell-usgs** [comment]
@dbuscombe-usgs did you try creating an updated environment and it failed?

**Comment by dnowacki-usgs** [comment]
I ended up creating my own. The harvest scripts are written for Python 2 so updated/current IOOS environments don't work. This is the output of `conda env export` from my faux IOOS2 environment.
```
name: IOOS2
channels:
  - conda-forge
  - defaults
dependencies:
  - _libgcc_mutex=0.1=main
  - asn1crypto=1.2.0=py27_0
  - bzip2=1.0.8=h516909a_1
  - ca-certificates=2019.9.11=hecc5488_0
  - certifi=2019.9.11=py27_0
  - cffi=1.13.2=py27h8022711_0
  - cftime=1.0.4.2=py27hc1659b7_0
  - chardet=3.0.4=py27_1003
  - cryptography=2.7=py27h72c5cf5_0
  - curl=7.65.3=hf8cf82a_0
  - enum34=1.1.6=py27_1001
  - hdf4=4.2.13=h9a582f1_1002
  - hdf5=1.10.5=nompi_h3c11f04_1104
  - icu=64.2=he1b5a44_1
  - idna=2.8=py27_1000
  - ipaddress=1.0.23=py_0
  - jpeg=9c=h14c3975_1001
  - krb5=1.16.3=h05b26f9_1001
  - libblas=3.8.0=14_openblas
  - libcblas=3.8.0=14_openblas
  - libcurl=7.65.3=hda55be3_0
  - libedit=3.1.20170329=hf8c457e_1001
  - libffi=3.2.1=he1b5a44_1006
  - libgcc-ng=9.1.0=hdf63c60_0
  - libgfortran-ng=7.3.0=hdf63c60_2
  - libiconv=1.15=h516909a_1005
  - liblapack=3.8.0=14_openblas
  - libnetcdf=4.7.1=nompi_h94020b1_101
  - libopenblas=0.3.7=h6e990d7_2
  - libssh2=1.8.2=h22169c7_2
  - libstdcxx-ng=9.1.0=hdf63c60_0
  - libxml2=2.9.10=hee79883_0
  - libxslt=1.1.33=h31b3aaa_0
  - lxml=4.4.1=py27h7ec2d77_0
  - ncurses=6.1=hf484d3e_1002
  - netcdf4=1.5.3=py27hf3213d6_0
  - numpy=1.16.5=py27h95a1406_0
  - openssl=1.1.1c=h516909a_0
  - pip=19.3.1=py27_0
  - pycparser=2.19=py27_1
  - pyopenssl=19.0.0=py27_0
  - pysgrid=0.3.5=py_0
  - pysocks=1.7.1=py27_0
  - python=2.7.15=h5a48372_1009
  - python-dateutil=2.8.1=py_0
  - pytz=2019.3=py_0
  - pyugrid=0.3.1=py_2
  - readline=8.0=hf8c457e_0
  - requests=2.22.0=py27_1
  - setuptools=41.6.0=py27_1
  - six=1.12.0=py27_1001
  - sqlite=3.30.1=hcee41ef_0
  - thredds_crawler=1.5.4=py_1
  - tk=8.6.9=hed695b0_1003
  - urllib3=1.25.6=py27_0
  - wheel=0.33.6=py27_0
  - xz=5.2.4=h14c3975_1001
  - zlib=1.2.11=h516909a_1006
prefix: /home/dnowacki/bin/miniconda3/envs/IOOS2
```

**Comment by rsignell-usgs** [comment]
@dnowacki-usgs, I see you have made progress:
```python
from owslib.csw import CatalogueServiceWeb
from owslib import fes
from pprint import pprint

endpoint = 'https://gamone.whoi.edu/csw'
csw = CatalogueServiceWeb(endpoint,timeout=60)

val = 'CMG_Portal'
filter1 = fes.PropertyIsLike(propertyname='apiso:AnyText',literal=('*%s*' % val),
                        escapeChar='\\',wildCard='*',singleChar='?')

filter_list = [filter1]

get_csw_records(csw, filter_list,  maxrecords=10000)
#csw.getrecords2(constraints=filter_list, maxrecords=1000)
print(len(csw.records.keys()))
```
returns:
```
2509
Wall time: 1min 24s
```

**Comment by dnowacki-usgs** [comment]
😄 

Progress, indeed. The issue now is getting pycsw to play nice with the daily/hourly cron jobs. With 2000+ records, the wipe process fails because of the way the sql query is set up. This is an [open issue over at pycsw](https://github.com/geopython/pycsw/issues/599) and persists with the new version of pycsw we are running. We could do our own little homebrew sqlite command to get around this, but...

Even after manually wiping the database, pycsw is so slow on gamone that it is almost unusable. Just setting up a database with `python bin/pycsw-admin.py -c setup_db -f default.cfg`  takes several minutes to complete, and that's to generate an empty database 250 kB in size.

Loading in the 2000+ static timeseries records with `python bin/pycsw-admin.py -c load_records -r -f default.cfg -p /store` takes _several hours_. Sqlite is generally highly performant, and even after everything is loaded the db is only about 280 MB in size, so things should not be taking this long. But since it takes many hours to complete, doing the wipe every night and reloading everything isn't really an option at this level of performance.

I don't think wiping the whole db should really be necessary, though. Force loading the dynamic records with `python bin/pycsw-admin.py -c load_records -r -f default.cfg -p /force -y` _should_ update already-existing records without requiring a full db wipe. However, I'm getting the same `sqlite3.IntegrityError: UNIQUE constraint failed: records.identifier` error that [you were getting 5 years ago](https://github.com/geopython/pycsw/issues/290) and appears long ago to have been fixed. 

**TL;DR** There is some bizarre behavior from pycsw, but I'm also starting to think there is something going awry with gamone. I've noticed simple things like `:wq`-ing a shell script in vi sometimes takes several seconds, and something like `mv cite.db cite-bak.db` can take seconds; both of these operations should be near instant. Filesystem response sometimes slows to a crawl, and seems especially bad when loading the records (though this could be a red herring). It's hard to imagine the whole system being I/O bound by updating this dinky little sqlite database, and with 40 cores/500 GB RAM and a single-thread python script we certainly aren't CPU/RAM bound. So I don't know what's going on. Have you encountered this?

**Comment by rsignell-usgs** [comment]
@tomkralidis, any sage advice here?

**Comment by rsignell-usgs** [comment]
@dnowacki-usgs , @kwilcox reports that using postgres, he is able to load all our ISO records in about 2 minutes.  Let's try that out.    gamone can't run docker-compose version 3, but let's see if this works with version 2:  https://github.com/axiom-data-science/docker-pycsw#production

**Comment by dnowacki-usgs** [comment]
Niiice. We appear to be in good shape now. Here's what I did.

Updated docker-compose.yml to version 2.1, which is the latest that gamone's docker supports. This update is required for the `depends_on` flag for postgres-enabled pycsw. Everything is now wrapped in the `services` key. This might result in a rebuild if you bounce a running container (I think this should only happen once). 

The relevant lines, which I adapted from the [docker-pycsw](https://github.com/axiom-data-science/docker-pycsw) readme:

```
  ###
  # pycsw: pycsw with postgres support
  ###
  pycsw:
    image: axiom/docker-pycsw:latest
    ports:
      - "8000:8000"
    container_name: pycsw
    depends_on:
      - postgresql
    volumes:
      - /opt/docker/pycsw/usgs-cmg-postgres.cfg:/opt/pycsw/default.cfg
      - /opt/docker/pycsw/force:/force
      - /opt/docker/pycsw/export:/export
      - /opt/docker/pycsw/store:/store
  
  ###
  # postgresql: postgres db for pycsw
  ###
  postgresql:
    image: kartoza/postgis:11.0-2.5
    stop_signal: SIGINT
    container_name: postgresql
    environment:
      POSTGRES_DBNAME: pycsw
      POSTGRES_USER: pycsw
      POSTGRES_PASS: pycsw
    volumes:
      - /opt/docker/postgresql:/var/lib/postgresql
```
And a change to the pycsw config file: 
```
database=postgresql://pycsw:pycsw@postgresql/pycsw
```
Loading the files from /store takes only a couple of minutes using Postgres. 🎉 

And searching for CMG_Portal results in a whole heap of files returned.

I archived the old container to `/opt/docker/pycsw-sqlite`.

**Comment by tomkralidis** [comment]
Glad to hear this worked out!

**Comment by kwilcox** [comment]
Are we good to go then @dnowacki-usgs?

**Comment by dnowacki-usgs** [comment]
Kyle, yes we are good to go. I'm going to close this.
